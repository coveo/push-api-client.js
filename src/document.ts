import {PermissionLevel, PermissionSet} from '@coveo/platform-client';

/**
 * A simple permission model. See https://docs.coveo.com/en/107 for more information.
 */
// export interface PermissionIdentityModel {
//   /**
//    * The name of the security identity.
//    *
//    * Examples:
//    *  - `asmith@example.com`
//    *  - `SampleTeam2`
//    */
//   identity: string;
//   /**
//    * The type of the identity.
//    * Valid values:
//    * - `UNKNOWN`
//    * - `USER` : Defines a single user.
//    * - `GROUP` : Defines an existing group of identities within the indexed system. Individual members of this group can be of any valid identity Type (USER, GROUP, or VIRTUAL_GROUP).
//    * - `VIRTUAL_GROUP` : Defines a group that doesn't exist within the indexed system. Mechanically, a `VIRTUAL_GROUP` is identical to a `GROUP`.
//    */
//   identityType: SinglePermissionIdentityType;
//   /**
//    * The security identity provider through which the security identity is updated.
//    *
//    * Defaults to the first security identity provider associated with the target Push source.
//    */
//   securityProvider?: string;
// }

type BaseMetadataValue = string | string[] | number | number[] | boolean;
type DictionaryMetadataValue = Record<string, BaseMetadataValue>;
export type MetadataValue = BaseMetadataValue | DictionaryMetadataValue;
export type Metadata = Record<string, MetadataValue>;

/**
 * The compression type that was applied to your compressed document.
 */
export type CompressionType =
  | 'UNCOMPRESSED'
  | 'DEFLATE'
  | 'GZIP'
  | 'LZMA'
  | 'ZLIB';

/**
 * A Coveo document.
 */
export interface Document {
  /**
   * The Uniform Resource Identifier (URI) that uniquely identifies the document in a Coveo index.
   *
   * Examples:
   * - `http://www.example.com/`
   * - `file://folder/text.txt`
   */
  uri: string;
  /**
   * The title of the document.
   */
  title: string;
  /**
   * The clickable URI associated with the document.
   */
  clickableUri?: string;
  /**
   * The author of the document.
   */
  author?: string;
  /**
   * The date of the document, represented as an ISO string.
   *
   * Optional, will default to indexation date.
   */
  date?: string;
  /**
   * The modified date of the document, represented as an ISO string.
   *
   * Optional, will default to indexation date.
   */
  modifiedDate?: string;
  /**
   * The permanent identifier of a document that does not change over time.
   *
   * Optional, will be derived from the document URI.
   */
  permanentId?: string;
  /**
   * The unique identifier (URI) of the parent item.
   *
   * Specifying a value for this key creates a relationship between the attachment item (child) and its parent item.
   *
   * This value also ensures that a parent and all of its attachments will be routed in the same index slice.
   */
  parentId?: string;
  /**
   * The textual (non-binary) content of the item.
   *
   * Whenever you're pushing a compressed binary item (such as XML/HTML, PDF, Word, or binary), you should use the CompressedBinaryData or CompressedBinaryDataFileId attribute instead, depending on the content size.
   *
   * Accepts 5 MB or less of uncompressed textual data.
   *
   * See https://docs.coveo.com/en/73 for more information.
   *
   * Example: `This is a simple string that will be used for searchability as well as to generate excerpt and summaries for the document.`
   */
  data?: string;
  /**
   * The original binary item content, compressed using one of the supported compression types (Deflate, GZip, LZMA, Uncompressed, or ZLib), and then Base64 encoded.
   *
   * You can use this parameter when you're pushing a compressed binary item (such as XML/HTML, PDF, Word, or binary) whose size is less than 5 MB.
   *
   * Whenever you're pushing an item whose size is 5 MB or more, use the CompressedBinaryDataFileIdproperty instead.
   *
   * If you're pushing less than 5 MB of textual (non-binary) content, you can use the data property instead.
   *
   * See https://docs.coveo.com/en/73 for more information.
   */
  compressedBinaryData?: {
    /**
     * The compression type that was applied to your document.
     */
    compressionType: CompressionType;
    /**
     * The base64 encoded binary data.
     *
     * Example: `eJxzrUjMLchJBQAK4ALN`
     */
    data: string;
  };
  /**
   * The metadata key-value pairs for a given document.
   *
   * Each metadata in the document must be unique.
   *
   * Metadata are case-insensitive (e.g., the Push API considers mykey, MyKey, myKey, MYKEY, etc. as identical).
   *
   * See https://docs.coveo.com/en/115 for more information.
   */
  metadata?: Metadata;
  /**
   * The list of permission sets for this item.
   *
   * This is useful when item based security is required (i.e., when security isn't configured at the source level).
   *
   * See https://docs.coveo.com/en/107 for more information.
   */
  permissions?: Array<PermissionSet | PermissionLevel>;
  /**
   * The file extension of the data you're pushing.
   *
   * This is useful when pushing a compressed item. The converter uses this information to identify how to correctly process the item.
   *
   * Values must include the preceding . character.
   *
   * Example: `.html`
   */
  fileExtension?: string;
}
