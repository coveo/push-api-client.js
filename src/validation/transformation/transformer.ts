// eslint-disable-next-line @typescript-eslint/no-unused-vars
import {BuiltInTransformers} from './builtInTransformers';

/**
 * Utility class to transform a string using a transformation function.
 *
 * For a list of built-in transformers, use {@link BuiltInTransformers}.
 */
export class Transformer {
  /**
   * @param {(str: string) => string} func The transformation function
   */
  public constructor(private func: (str: string) => string) {}

  /**
   * Executes the transformation function to the input string
   *
   * @param {string} str The string to transform
   * @return {*} The transformed string
   */
  public transform(str: string) {
    return this.func(str);
  }
}
