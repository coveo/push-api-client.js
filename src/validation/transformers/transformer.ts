export type Transformer = (str: string) => string;

class InternalBuiltInTransformers {
  /**
   * A transformer that removes leading numbers and non-alphanumeric characters.
   *
   * _Transformation examples:_
   *
   * - `Field-With=Special+Characters` => `fieldwithspecialcharacters`
   * - `01-Numeric-Field-#01` => `numericfield01`
   */
  public get toLowerCase(): Transformer {
    return (field: string) =>
      field
        .toLowerCase()
        // Remove non-alpha numeric
        .replace(/[^a-z0-9]/gi, '')
        // Remove any leading number
        .replace(/^[0-9]+/g, '');
  }
  /**
   * A transformer that removes leading numbers and replaces every non-alphanumeric characters with underscores
   *
   * _Transformation examples:_
   *
   * - `Field-With=Special+Characters` => `field_with_special_characters`
   * - `01-Numeric-Field-#01` => `numeric_field_01`
   */
  public get toSnakeCase(): Transformer {
    return (field: string) =>
      field
        .toLowerCase()
        // Replace non-alpha numeric with underscores
        .replace(/[^a-z0-9]/gi, '_')
        // Remove any leading number or underscore
        .replace(/^[0-9_]+/g, '')
        // Remove consecutive underscores
        .replace(/_{2,}/g, '_')
        // Remove any trailing underscore
        .replace(/_$/g, '');
  }
  /**
   * A transformer that does nothing but return the string supplied to it.
   * Good as a default transformer function.
   *
   * _Transformation examples:_
   *
   * - `foo` => `foo`
   * - `bar` => `bar`
   */
  public get identity(): Transformer {
    return (field: string) => field;
  }
}

export const BuiltInTransformers = new InternalBuiltInTransformers();
