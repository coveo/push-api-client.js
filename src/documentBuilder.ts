import dayjs = require('dayjs');
import {createHash} from 'crypto';
import {CompressionType, Document, Metadata, MetadataValue} from './document';
import {SecurityIdentityBuilder} from './securityIdentityBuilder';

/**
 * Utility class to build a {@link Document}.
 */
export class DocumentBuilder {
  private doc: Document;
  /**
   *
   * @param uri The URI of the document. See {@link Document.uri}
   * @param title The title of the document. See {@link Document.title}
   */
  constructor(private uri: string, title: string) {
    this.doc = {
      uri,
      title,
      metadata: {},
      permissions: {allowAnonymous: false},
    };
  }

  /**
   * Set the data of the document. See {@link Document.data}
   * @param data
   * @returns
   */
  public withData(data: string) {
    this.doc.data = data;
    return this;
  }

  /**
   * Set the date of the document. See {@link Document.date}
   * @param date
   * @returns
   */
  public withDate(date: Date | string | number) {
    this.doc.date = this.validateDateAndReturnValidDate(date);
    return this;
  }

  /**
   * Set the modified date of the document. See {@link Document.modifiedDate}
   * @param date
   * @returns
   */
  public withModifiedDate(date: Date | string | number) {
    this.doc.modifiedDate = this.validateDateAndReturnValidDate(date);
    return this;
  }

  /**
   * Set the permanentID of the document. See {@link Document.permanentId}
   * @param permanentID
   * @returns
   */
  public withPermanentId(permanentID: string) {
    this.doc.permanentId = permanentID;
    return this;
  }

  /**
   * Set the base64 encoded, compressed binary data of the document. See {@link Document.compressedBinaryData}
   * @param data
   * @param compressionType
   * @returns
   */
  public withCompressedBinaryData(
    data: string,
    compressionType: CompressionType
  ) {
    this.validateCompressedBinaryData(data);
    this.doc.compressedBinaryData = {
      data,
      compressionType,
    };
    return this;
  }

  /**
   * Set the file extension on the document. See {@link Document.fileExtension}
   * @param extension
   * @returns
   */
  public withFileExtension(extension: string) {
    //TODO: Validate valid file extension
    this.doc.fileExtension = extension;
    return this;
  }

  /**
   * Set the parentID on the document. See {@link Document.parentId}
   * @param id
   * @returns
   */
  public withParentID(id: string) {
    this.doc.parentId = id;
    return this;
  }

  /**
   * Set the clickableURI on the document. See {@link Document.clickableUri}
   * @param clickURI
   * @returns
   */
  public withClickableUri(clickURI: string) {
    // TODO: Validate that it's a valid link/proper scheme... ?
    // Unsure
    this.doc.clickableUri = clickURI;
    return this;
  }

  /**
   * Set the author on the document. See {@link Document.author}
   * @param author
   * @returns
   */
  public withAuthor(author: string) {
    this.doc.author = author;
    return this;
  }

  /**
   * Add a single metadata key and value pair on the document. See {@link Document.metadata}
   * @param key
   * @param value
   * @returns
   */
  public withMetadataValue(key: string, value: MetadataValue) {
    // TODO: validate reserved names
    this.doc.metadata![key] = value;
    return this;
  }

  /**
   * Set metadata on the document. See {@link Document.metadata}
   * @param metadata
   * @returns
   */
  public withMetadata(metadata: Metadata) {
    for (const [k, v] of Object.entries(metadata)) {
      this.withMetadataValue(k, v);
    }
    return this;
  }

  /**
   * Set allowed identities on the document. See {@link Document.permissions}
   * @param securityIdentityBuilder
   * @returns
   */
  public withAllowedPermissions(
    securityIdentityBuilder: SecurityIdentityBuilder
  ) {
    this.setPermission(securityIdentityBuilder, 'allowedPermissions');
    return this;
  }

  /**
   * Set denied identities on the document. See {@link Document.permissions}
   * @param securityIdentityBuilder
   * @returns
   */
  public withDeniedPermissions(
    securityIdentityBuilder: SecurityIdentityBuilder
  ) {
    this.setPermission(securityIdentityBuilder, 'deniedPermissions');
    return this;
  }

  /**
   * Set allowAnonymous for permissions on the document. See {@link Document.permissions}
   * @param allowAnonymous
   * @returns
   */
  public withAnonymousUsers(allowAnonymous: boolean) {
    this.doc.permissions!.allowAnonymous = allowAnonymous;
    return this;
  }

  public build() {
    return this.doc;
  }

  /**
   * Marshal the document into a JSON format accepted by the push API.
   * @returns
   */
  public marshal() {
    this.validateAndFillMissing();
    const out = {
      ...this.doc,
      ...this.marshalMetadata(),
      ...this.marshalCompressedBinaryData(),
      ...this.marshalPermissions(),
    };
    delete out.metadata;
    return out;
  }

  private marshalMetadata() {
    if (!this.doc.metadata) {
      return {};
    }
    const out: Metadata = {};
    for (const [k, v] of Object.entries(this.doc.metadata)) {
      out[k] = v;
    }
    return out;
  }

  private marshalCompressedBinaryData() {
    if (!this.doc.compressedBinaryData) {
      return {};
    }
    return {
      compressedBinaryData: this.doc.compressedBinaryData.data,
      compressionType: this.doc.compressedBinaryData.compressionType,
    };
  }

  private marshalPermissions() {
    if (!this.doc.permissions) {
      return '';
    }
    return {
      allowAnonymous: this.doc.permissions.allowAnonymous,
      allowedPermissions: this.doc.permissions.allowedPermissions?.map((p) =>
        JSON.stringify(p)
      ),
      deniedPermissions: this.doc.permissions.deniedPermissions?.map((p) =>
        JSON.stringify(p)
      ),
    };
  }

  private validateAndFillMissing() {
    // TODO: validation that cannot be performed on a single property, but requires looking at multiple property at the same time.
    // or validate that we don't have both `data` AND `compressedBinaryData`.
    // Could also use https://www.npmjs.com/package/@coveo/bueno to validate schema (useful for pure JS users).
    if (!this.doc.permanentId) {
      this.doc.permanentId = this.generatePermanentId();
    }
    return;
  }

  private generatePermanentId() {
    return createHash('sha256').update(this.doc.uri).digest('hex');
  }

  private validateDateAndReturnValidDate(d: Date | string | number) {
    const validatedDate = dayjs(d);
    return validatedDate.toISOString();
  }

  private validateCompressedBinaryData(data: string) {
    const isBase64 = Buffer.from(data, 'base64').toString('base64') === data;
    if (!isBase64) {
      throw 'Invalid compressedBinaryData: When using compressedBinaryData, the data must be base64 encoded.';
    }
  }

  private setPermission(
    securityIdentityBuilder: SecurityIdentityBuilder,
    permissionSection: 'allowedPermissions' | 'deniedPermissions'
  ) {
    const identities = securityIdentityBuilder.build();
    if (Array.isArray(identities)) {
      this.doc.permissions![permissionSection] = identities;
    } else {
      this.doc.permissions![permissionSection] = [identities];
    }
  }
}
