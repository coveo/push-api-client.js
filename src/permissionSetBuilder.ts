import {PermissionSetModel} from './document';
import {SecurityIdentityBuilder} from './securityIdentityBuilder';

type PermissionSection = keyof Pick<
  PermissionSetModel,
  'allowedPermissions' | 'deniedPermissions'
>;

// TODO: test
export class PermissionSetBuilder {
  private permissionSet: PermissionSetModel;

  /**
   * Creates an instance of PermissionSetBuilder.
   * TODO: document on permission set
   * @param {boolean} allowAnonymous Set allowAnonymous for permissions on the document
   */
  public constructor(allowAnonymous: boolean) {
    this.permissionSet = {
      allowAnonymous,
      allowedPermissions: [],
      deniedPermissions: [],
    };
  }

  /**
   * Set allowed identities on the document. See {@link Document.permissions}
   * @param securityIdentityBuilder
   * @returns
   */
  public withAllowedPermissions(
    securityIdentityBuilder: SecurityIdentityBuilder
  ) {
    this.setPermission(securityIdentityBuilder, 'allowedPermissions');
    return this;
  }

  /**
   * Set denied identities on the document. See {@link Document.permissions}
   * @param securityIdentityBuilder
   * @returns
   */
  public withDeniedPermissions(
    securityIdentityBuilder: SecurityIdentityBuilder
  ) {
    this.setPermission(securityIdentityBuilder, 'deniedPermissions');
    return this;
  }

  /**
   * Build and return the security identity or identities.
   * @returns
   */
  public build(): PermissionSetModel {
    return this.permissionSet;
  }

  private setPermission(
    securityIdentityBuilder: SecurityIdentityBuilder,
    permissionSection: PermissionSection
  ) {
    const identities = securityIdentityBuilder.build();
    if (Array.isArray(identities)) {
      this.permissionSet[permissionSection] =
        this.permissionSet[permissionSection]?.concat(identities);
    } else {
      this.permissionSet[permissionSection]?.push(identities);
    }
  }
}
