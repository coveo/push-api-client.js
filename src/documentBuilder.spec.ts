import {PermissionSetModel} from './document';
import {DocumentBuilder} from './documentBuilder';
import {PermissionSetBuilder} from './permissionSetBuilder';
import {
  GroupSecurityIdentityBuilder,
  UserSecurityIdentityBuilder,
} from './securityIdentityBuilder';
import {BuiltInTransformers} from './validation/transformers/transformer';

describe('DocumentBuilder', () => {
  let docBuilder: DocumentBuilder;
  beforeEach(() => {
    docBuilder = new DocumentBuilder('uri', 'title');
  });

  it('should marshal title', () => {
    expect(docBuilder.marshal().title).toBe('title');
  });

  it('should marshal uri to documentId', () => {
    expect(docBuilder.marshal().documentId).toBe('uri');
  });

  it('should marshal data', () => {
    expect(docBuilder.withData('foo').marshal().data).toBe('foo');
  });

  it('should marshal date', () => {
    const d = new Date();
    expect(docBuilder.withDate(d).marshal().date).toMatch(d.toISOString());
  });

  it('should marshal author', () => {
    expect(docBuilder.withAuthor('bob').marshal().author).toBe('bob');
  });

  it('should marshal clickableUri', () => {
    expect(docBuilder.withClickableUri('click').marshal().clickableUri).toBe(
      'click'
    );
  });

  it('should marshal fileExtension', () => {
    expect(docBuilder.withFileExtension('.html').marshal().fileExtension).toBe(
      '.html'
    );
  });

  it('should marshal metadata', () => {
    expect(
      docBuilder
        .withMetadata({
          foo: 'bar',
          buzz: ['bazz', 'bozz'],
          baz: {'': 'defaultValue', key1: 'value1'},
        })
        .marshal()
    ).toMatchObject({
      foo: 'bar',
      buzz: ['bazz', 'bozz'],
      baz: {'': 'defaultValue', key1: 'value1'},
    });
  });

  it('should throw error for invalid single metadata value', () => {
    expect(() => {
      docBuilder.withMetadataValue('f-o=o', 'bar').marshal();
    }).toThrowError('f-o=o');
  });

  it('should throw error for invalid metadata values', () => {
    expect(() => {
      docBuilder
        .withMetadata({'f-o=o': 'bar', '<buzz>': ['bazz', 'bozz']})
        .marshal();
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw error for invalid metadata values even after transformation', () => {
    const poorTransformer = (text: string) => text.replace(/[\W_]+/g, '*');
    expect(() => {
      docBuilder
        .withMetadata(
          {'f-o=o': 'bar', '<buzz>': ['bazz', 'bozz']},
          poorTransformer
        )
        .marshal();
    }).toThrowErrorMatchingSnapshot();
  });

  describe.each([
    {
      title: 'default',
      transformer: BuiltInTransformers.toLowerCase,
      expectedKey: 'foo',
    },
    {
      title: 'underscore',
      transformer: BuiltInTransformers.toSnakeCase,
      expectedKey: 'f_o_o',
    },
  ])('when using the $title transformer', ({transformer, expectedKey}) => {
    it(`should format to ${expectedKey}`, () => {
      expect(
        new DocumentBuilder('uri', 'title')
          .withMetadataValue('f-o=o', 'bar', transformer)
          .marshal()
      ).toMatchObject({
        [expectedKey]: 'bar',
      });
    });
  });

  it('should marshal single metadata value', () => {
    expect(docBuilder.withMetadataValue('foo', 'bar').marshal()).toMatchObject({
      foo: 'bar',
    });
  });

  it('should not marshal metadata key', () => {
    expect(
      (
        docBuilder
          .withMetadata({foo: 'bar', buzz: ['bazz', 'bozz']})
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          .marshal() as any
      ).metadata
    ).toBeUndefined();
  });

  it('should marshal modifiedDate', () => {
    const d = new Date();
    expect(docBuilder.withModifiedDate(d).marshal().modifiedDate).toMatch(
      d.toISOString()
    );
  });

  it('should marshal parentId', () => {
    expect(docBuilder.withParentID('id').marshal().parentId).toBe('id');
  });

  it('should marshal permanentId', () => {
    expect(docBuilder.withPermanentId('id').marshal().permanentId).toBe('id');
  });

  it('should generate permanentid', () => {
    docBuilder = new DocumentBuilder('https://foo.com', 'bar');
    expect(docBuilder.marshal().permanentId).toBe(
      'aa2e0510b66edff7f05e2b30d4f1b3a4b5481c06b69f41751c54675c5afb'
    );
  });

  it('throws when adding a reserved key name metadata', () => {
    const theseShouldThrow = [
      'compressedBinaryData',
      'compressedBinaryDataFileId',
      'parentId',
      'fileExtension',
      'data',
      'permissions',
      'documentId',
      'orderingId',
    ];

    for (const shouldThrow of theseShouldThrow) {
      expect(() => docBuilder.withMetadataValue(shouldThrow, 'foo')).toThrow();
    }
  });

  it('should validate file extension', () => {
    expect(() => docBuilder.withFileExtension('nope')).toThrow();
  });

  // TODO: describe permission sets

  it('should marshal allowedPermissions', () => {
    const permissionSet = new PermissionSetBuilder(
      false
    ).withAllowedPermissions(new UserSecurityIdentityBuilder('bob@foo.com'));

    expect(
      docBuilder.withPermissionSet(permissionSet).marshal().permissions[0]
    ).toMatchObject({
      allowedPermissions: expect.arrayContaining([
        expect.objectContaining({identity: 'bob@foo.com'}),
      ]),
    });
  });

  it('should marshal allowedPermissions in multiple #withAllowedPermissions', () => {
    const permissionSet = new PermissionSetBuilder(false)
      .withAllowedPermissions(new UserSecurityIdentityBuilder('bob@foo.com'))
      .withAllowedPermissions(new GroupSecurityIdentityBuilder('my_group'));

    expect(
      docBuilder.withPermissionSet(permissionSet).marshal().permissions[0]
    ).toMatchObject({
      allowedPermissions: expect.arrayContaining([
        expect.objectContaining({
          identity: 'bob@foo.com',
          identityType: 'USER',
        }),
        expect.objectContaining({
          identity: 'my_group',
          identityType: 'GROUP',
        }),
      ]),
    });
  });

  it('should marshal multiple allowedPermissions', () => {
    const permissionSet = new PermissionSetBuilder(
      false
    ).withAllowedPermissions(
      new UserSecurityIdentityBuilder(['bob@foo.com', 'sue@foo.com'])
    );

    expect(
      docBuilder.withPermissionSet(permissionSet).marshal().permissions[0]
    ).toMatchObject({
      allowedPermissions: expect.arrayContaining([
        expect.objectContaining({identity: 'bob@foo.com'}),
        expect.objectContaining({identity: 'sue@foo.com'}),
      ]),
    });
  });

  it('should marshal deniedPermissions', () => {
    const permissionSet = new PermissionSetBuilder(false).withDeniedPermissions(
      new UserSecurityIdentityBuilder('bob@foo.com')
    );

    expect(
      docBuilder.withPermissionSet(permissionSet).marshal().permissions[0]
    ).toMatchObject({
      deniedPermissions: expect.arrayContaining([
        expect.objectContaining({identity: 'bob@foo.com'}),
      ]),
    });
  });

  it('should marshal deniedPermissions in multiple #withDeniedPermissions', () => {
    const permissionSet = new PermissionSetBuilder(false)
      .withDeniedPermissions(new UserSecurityIdentityBuilder('bob@foo.com'))
      .withDeniedPermissions(new GroupSecurityIdentityBuilder('my_group'));

    expect(
      docBuilder.withPermissionSet(permissionSet).marshal().permissions[0]
    ).toMatchObject({
      deniedPermissions: expect.arrayContaining([
        expect.objectContaining({
          identity: 'bob@foo.com',
          identityType: 'USER',
        }),
        expect.objectContaining({
          identity: 'my_group',
          identityType: 'GROUP',
        }),
      ]),
    });
  });

  it('should marshal multiple deniedPermissions', () => {
    const permissionSet = new PermissionSetBuilder(false).withDeniedPermissions(
      new UserSecurityIdentityBuilder(['bob@foo.com', 'sue@foo.com'])
    );

    expect(
      docBuilder.withPermissionSet(permissionSet).marshal().permissions[0]
    ).toMatchObject({
      deniedPermissions: expect.arrayContaining([
        expect.objectContaining({identity: 'bob@foo.com'}),
        expect.objectContaining({identity: 'sue@foo.com'}),
      ]),
    });
  });

  // it('should marshal allowedPermissions to an empty array when undefined', () => {
  //   const permissionSet = docBuilder.marshal()
  //     .permissions[0] as PermissionSetModel;
  //   expect(permissionSet.allowedPermissions!.length).toBe(0);
  // });

  // it('should marshal deniedPermissions to an empty array when undefined', () => {
  //   const permissionSet = docBuilder.marshal()
  //     .permissions[0] as PermissionSetModel;
  //   expect(permissionSet.deniedPermissions!.length).toBe(0);
  // });
  it('should marshal permissions to an empty array when undefined', () => {
    const permissionSet = docBuilder.marshal().permissions;
    expect(permissionSet).toEqual([]);
  });

  // TODO: test with levels
  // TODO: add sample file with permission levels
});
