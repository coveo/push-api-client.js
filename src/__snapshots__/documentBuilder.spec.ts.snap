// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentBuilder should throw error for invalid metadata values 1`] = `
"
The following field names are still invalid after transformation:
  • f-o=o --> f-o=o
  • <buzz> --> <buzz>

Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. Field names must be at least one character long and must start with a lowercase letter.
    "
`;

exports[`DocumentBuilder should throw error for invalid metadata values even after transformation 1`] = `
"
The following field names are still invalid after transformation:
  • f-o=o --> f*o*o
  • <buzz> --> *buzz*

Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. Field names must be at least one character long and must start with a lowercase letter.
    "
`;

exports[`DocumentBuilder when combining multiple permission sets should marshal permission level 1`] = `
[
  {
    "name": "level1",
    "permissionSets": [
      {
        "allowAnonymous": true,
        "allowedPermissions": [],
        "deniedPermissions": [
          {
            "identity": "asmith@example.com",
            "identityType": "User",
            "securityProvider": "Email Security Provider",
          },
        ],
      },
      {
        "allowAnonymous": false,
        "allowedPermissions": [
          {
            "identity": "SampleTeam1",
            "identityType": "Group",
          },
          {
            "identity": "emitchell@example.com",
            "identityType": "User",
            "securityProvider": "Email Security Provider",
          },
        ],
        "deniedPermissions": [],
      },
    ],
  },
  {
    "name": "level2",
    "permissionSets": [
      {
        "allowAnonymous": false,
        "allowedPermissions": [
          {
            "identity": "MysteryUserX",
            "identityType": "User",
            "securityProvider": "Email Security Provider",
          },
        ],
        "deniedPermissions": [
          {
            "identity": "SampleGroup",
            "identityType": "VirtualGroup",
          },
        ],
      },
    ],
  },
]
`;
